js/ImagePane.js

```diff
@@ -7,88 +7,41 @@
  *
  */
 
-import React from 'react';
+import React, { useEffect, useState, useRef } from 'react';
 import EventSystem from './EventSystem';
 import Pane from './Pane';
 
 const DEFAULT_HEIGHT = 400;
 const DEFAULT_WIDTH = 300;
 
-class ImagePane extends React.Component {
-  _paneRef = null;
-  _imgRef = null;
-  _natHeight = null;
-  _natWidth = null;
-
-  state = {
-    lastModTime: 0,
-    scale: 1,
-    tx: 0,
-    ty: 0,
-    selected: this.props.selected,
-    mouse_location: { x: 0, y: 0, visibility: 'hidden' },
-  };
-
-  drag_start_x = null;
-  drag_start_y = null;
-
-  componentWillReceiveProps = (nextProps) => {
-    if (nextProps.selected !== this.props.selected) {
-      this.setState({ selected: nextProps.selected });
-    }
-    if (
-      (nextProps.width != this.props.width ||
-        nextProps.height != this.props.height) &&
-      Math.abs(this.state.scale - 1) > Number.EPSILON
-    ) {
-      // Reset the image settings when the user resizes the window. Avoid
-      // constantly resetting the zoom level when user has not zoomed.
-      this.handleReset();
-    }
-  };
-
-  onEvent = (event) => {
-    if (!this.props.isFocused) {
-      return;
-    }
-
-    switch (event.type) {
-      case 'keydown':
-      case 'keypress':
-        event.preventDefault();
-        break;
-      case 'keyup':
-        this.props.appApi.sendPaneMessage({
-          event_type: 'KeyPress',
-          key: event.key,
-          key_code: event.keyCode,
-        });
-        break;
-      case 'click':
-        this.props.appApi.sendPaneMessage({
-          event_type: 'Click',
-          image_coord: this.state.mouse_location,
-        });
-        break;
-    }
-  };
-
-  componentDidMount() {
-    EventSystem.subscribe('global.event', this.onEvent);
-  }
-
-  componentWillUnmount() {
-    EventSystem.unsubscribe('global.event', this.onEvent);
-  }
-
-  handleDownload = () => {
+function ImagePane(props) {
```

Comment:
As a _tiny_ nit, a common react pattern when we know that _this_ component uses specific `props` and the downstream component (that we extend to with `{...props}`) uses different ones is to do parameter de-structuring inline, like:
```
function ImagePane({ known, used, values, ...props})
```
This both clarifies values that this component needs as well as prevents them from being passed down.

In this particular class though, there's a lot of stuff that is being used here and still passed down to the pane, so for these I like doing a de-structure right at the top:
```
const {appApi, height, width, other, args, that, are, needed} = props;
```
This still makes it easy to reason about the required prop arguments upfront.

Reply:
Thanks for pointing this out, I did not know that syntax, yet! :)

**I have adapted all Components to include such a de-construction of all prop-values (if possible) at the start of the function.**

I have added that change as a new commit for easier inspection of the changes.
(If you want, I can squash these changes also into the respective previous commits before merging).